persistent/cPersistence.c

```diff
@@ -552,8 +554,20 @@ pickle___setstate__(PyObject *self, PyObject *state)
         }
 
         PyDict_Clear(*dict);
-        if (PyDict_Update(*dict, state) < 0)
-            return NULL;
+
+        i = 0;
+        while (PyDict_Next(state, &i, &d_key, &d_value)) {
+            /* normally the keys for instance attributes are
+               interned.  we should try to do that here. */
+            Py_INCREF(d_key);
```

Comment:
Nitpick:  I'd rather have the `Py_INCREF` and `Py_DECREF` wrapped around the `INTERN_INPLACE` call (the thing which could actually decref the key), and remove the need for the extra decref in the failure case for `PyObject_SetItem`.